// shua/dist/utils/logger.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
var Logger = /** @class */ (function () {
    function Logger() {
    }
    return Logger;
}());
exports.Logger = Logger;
var ConsoleLogger = /** @class */ (function (_super) {
    __extends(ConsoleLogger, _super);
    function ConsoleLogger() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        _this.debugFlag = false;
        return _this;
    }
    ConsoleLogger.prototype.debug = function (message) {
        if (!this.debugFlag) {
            return;
        }
        console.debug("" + message);
    };
    ConsoleLogger.prototype.info = function (message) {
        console.info("" + message);
    };
    ConsoleLogger.prototype.warning = function (message) {
        console.warn("" + message);
    };
    ConsoleLogger.prototype.error = function (message, error) {
        console.error("" + message);
        if (error) {
            console.log(error);
        }
    };
    ConsoleLogger.prototype.enableDebug = function () {
        this.debugFlag = true;
    };
    return ConsoleLogger;
}(Logger));
exports.ConsoleLogger = ConsoleLogger;
exports.default = new ConsoleLogger();
